File: DLL/DllCommunication.cpp
/*
* OSPE - Open Source Packet Editor
* Copyright(C) 2018-2019 Javier Pereda <https://github.com/elecyb>
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "stdafx.h"
#include "DllCommunication.h"
#include "OspeDll.h"
#include "FilterManager.h"
#include <sstream>
#include "Utils.h"

//#define DBG_BLOCKPKT // para debuggear packets bloqueados
//#define DBG_SENDPACKETDATA // para debuggear data de los packets 

char readData[IPC_BLOCK_SIZE];



// Agrega a info la informacion del socket (ip, port) y tambien el tipo de funcion y size del packet
void SetInfo(SOCKET socket, Functions functionId, int length, PacketInfo* info)
{
	//char ipstr[INET6_ADDRSTRLEN];
	BYTE ipver = IPV4;

	struct sockaddr_storage addrLocal;
	struct sockaddr_storage addrRemote;
	socklen_t addr_len = sizeof(struct sockaddr_storage);

	if (getsockname(socket, (struct sockaddr*)&addrLocal, &addr_len) == SOCKET_ERROR)
		Utils::errorLog(Utils::strm(2, "getsockname: ", Utils::IntToString(WSAGetLastError())));

	if (getpeername(socket, (struct sockaddr*)&addrRemote, &addr_len) == SOCKET_ERROR)
		Utils::errorLog(Utils::strm(2, "getpeername: ", Utils::IntToString(WSAGetLastError())));

	// deal with both IPv4 and IPv6:
	if (addrLocal.ss_family == AF_INET)
	{
		info->localIp = htonl(((struct sockaddr_in *)&addrLocal)->sin_addr.s_addr);
		info->localPort = htons(((struct sockaddr_in *)&addrLocal)->sin_port);

		info->remoteIp = htonl(((struct sockaddr_in *)&addrRemote)->sin_addr.s_addr);
		info->remotePort = htons(((struct sockaddr_in *)&addrRemote)->sin_port);

	}
	else
	{ // AF_INET6
		//Falta implementar
		Utils::errorLog("CRASH ALERT: IPv6 SOCKET NOT IMPLEMENTED");
		//struct sockaddr_in6 *sock_addr = (struct sockaddr_in6 *)&addrFrom;
		//info->portFrom = ntohs(sock_addr->sin6_port);
		//inet_ntop(AF_INET6, &sock_addr->sin6_addr, ipstr, sizeof ipstr);
		ipver = IPV6;
	}

	//Envía la versión del protocolo a través del campo functionId
	info->functionId = (Functions)(functionId | ipver);
	info->socketId = socket;
	info->size = length;
}

// This method writes the packet to the memory mapped file then returns
void ProcessPacket(Functions functionId, char* &buffer, int &length, SOCKET socket, bool &blocked)
{
	char * name = (char*)malloc(IPC_MAX_NAME);
	sprintf(name, "%s%u", "OSPEPACKETBUFF", GetCurrentProcessId());
	static osIPC::Client client(name);


	if (!client.IsOk())
	{
#ifdef DBG_SENDPACKETDATA 
		Utils::errorLog("Fail to open MMF!!", 1);
#endif
		return;
	}		

	// Set needed packet info (ip, port, Function, Size)
	PacketInfo info;
	if (socket != NULL)
		SetInfo(socket, functionId, (int)length, &info);
	else {
		info.functionId = functionId;
		info.localIp = 0;
		info.localPort = 0;
		info.remoteIp = 0;
		info.remotePort = 0;
		info.size = length;
		info.socketId = 0;
	}

	// Set data


#ifdef DBG_SENDPACKETDATA
	std::stringstream sinfo;
	sinfo << "GOT Info - FunctionID=" << info.functionId << " LocalIp=" << info.localIp << " LocalPort=" << info.localPort << " RemoteIp="
		<< info.remoteIp << " RemotePort=" << info.remotePort << " Size=" << info.size;
	Utils::errorLog((char*)sinfo.str().c_str(), 2);
#endif

	char* pBuff = (char *) malloc(sizeof(info) + length);
	memcpy(pBuff, &info, sizeof(info));
	memcpy(pBuff + sizeof(info), buffer, length);

	bool breakpoint = false;
	if (isFiltering)
		// Only filter if it wasn't blocked, makes no sense filter a blocked packet
		if (!CheckPacketBlock(pBuff + sizeof(info), info.size, (FilterCaptureFuncs)info.functionId))
			// If filter is a Breakpoint filter then stop filtering
			if (!CheckPacketBreak(pBuff + sizeof(info), info.size, (FilterCaptureFuncs)info.functionId)) {
				if (DoFilteringForPacket(pBuff + sizeof(info), info.size, (FilterCaptureFuncs)info.functionId))
					// Copy new filtered data
					memcpy((void*)buffer, pBuff + sizeof(info), length);
			}
			else 
				breakpoint = true;							
		else
			blocked = true;

	client.write(pBuff, sizeof(info) + length); // Send the packet to ospe

	if (breakpoint)
	{
		hMainThread = OpenThread(THREAD_ALL_ACCESS, FALSE, GetCurrentThreadId());
		SuspendThread(hMainThread); // Suspend thread until we receive new data from ospe
		UINT16 newLength = 0; 
		newLength = ((UINT8)readData[2] << 8) | ((UINT8)readData[1]); // Set the length for the new packet
		length = newLength;
		buffer = (char *)malloc(length);
		memcpy((void*)buffer, &readData[3], length); // read new data
	}
	
	free(pBuff);
}

void ProcessPacket(Functions functionId, char*& buffer, int& length) 
{
	bool blocked = false;
	ProcessPacket(functionId, buffer, length, NULL, blocked);
}

void InjectPacket() 
{
	SOCKET s = (((UINT8)readData[2] << 8) | (UINT8)readData[1]);
	int len = (((UINT8)readData[4] << 8) | (UINT8)readData[3]);
	const char * buf = &readData[5];
	int flags = MSG_DONTROUTE;
	send(s, buf, len, flags);
}

// Command reader thread worker
DWORD WINAPI Command_Reader(LPVOID context)
{
    // Continuously read data
    for (;;)
    {
        server.read(readData, IPC_BLOCK_SIZE, INFINITE);
		ServerCodes sc = (ServerCodes)readData[0];
		switch (sc)
		{
			case SCODE_INJECTPACKET:
				InjectPacket();
				break;
			case SCODE_SETPACKET:
				ResumeThread(hMainThread);
				break;
			case SCODE_STARTFILTERING:
				{		
					CovertBytesToFilterList(readData);
					isFiltering = true;
				}
				break;
			case SCODE_STOPFILTERING:
				isFiltering = false;
				break;
			case SCODE_STARTCAPTURE:
				isCapturing = true;
				break;
			case SCODE_STOPCAPTURE:
				isCapturing = false;
				break;
			case SCODE_LOADDLLEX:
				LoadDllEx(&readData[1]);
				break;
			case SCODE_UNLOADDLLEX:
				UnLoadDllEx();
				break;
			default:
				Utils::errorLog("UNKNOWN SERVER CODE!");
		}

    }

    // Success
    return 0;
};

File: DllCommunication.cs
/*
* OSPE - Open Source Packet Editor
* Copyright(C) 2018-2019 Javier Pereda <https://github.com/elecyb>
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Runtime.InteropServices;
using System.Diagnostics;
using System.IO;
using System.IO.MemoryMappedFiles;
using System.Security.AccessControl;
using SharedMemory;
using System.Threading;
using Buffer = System.Buffer;

namespace OSPE
{
    public enum ServerCodes : uint
    {
        SCODE_NOP,
        SCODE_INJECTPACKET,
        SCODE_SETPACKET,
        SCODE_STARTFILTERING,
        SCODE_STOPFILTERING,
        SCODE_STARTCAPTURE,
        SCODE_STOPCAPTURE,
        SCODE_LOADDLLEX,
        SCODE_UNLOADDLLEX,
    };

    static class DllCommunication
    {
        private static bool CheckMMFileExists(string fileName)
        {
            try
            {
                MemoryMappedFile.OpenExisting(fileName);
            }catch(FileNotFoundException)
            {
                return false;
            }
            return true;
        }

        public static void SetTargetPid(int pid)
        {
            _pktMmfName = @"OSPEPACKETBUFF" + pid;
            _cmdMmfName = @"OSPECMDBUFF" + pid;
        }


        // Memory Mapped File - Packet MMF data
        /*
                |       PacketInfo   |                      PacketInfo.Size                                |
                |        20-bytes    |------------------------------Data-----------------------------------|
                |-----------------------------------------_pktMmfBufferSize--------------------------------|
        */

        private static string _pktMmfName;
        private static int _pktMmfNodeCount = 50;
        private static int _pktMmfBufferSize = 32792;
        public static void StartPacketReaderMMF()
        {
            int nodeBufferSize = sizeof(byte) * _pktMmfBufferSize;
            SharedMemory.CircularBuffer packetMMF;
            if (!CheckMMFileExists(_pktMmfName))
                packetMMF = new SharedMemory.CircularBuffer(_pktMmfName, _pktMmfNodeCount, nodeBufferSize);
            else
            {
                packetMMF = new SharedMemory.CircularBuffer(_pktMmfName);
            }

            byte[] writtenData = new byte[_pktMmfBufferSize];

            //int threadCount = 0;
            Action reader = () =>
            {
                //int myThreadIndex = Interlocked.Increment(ref threadCount);
                //Output.outMsgBox("Started thread {0}: ", threadCount);

                for (; ; )
                {
                    //var t1 = Environment.TickCount; Debug.WriteLine("packetMMF.Read(writtenData) {0}", ID);
                    //var t2 = Environment.TickCount; if (t2-t1 > 0) Debug.WriteLine("ANTES de packetMMF.Read {0} - diff: {1}", ID, t2-t1);

                    //The number of milliseconds to wait, or Timeout.Infinite (-1) to wait indefinitely. 
                    int amount = packetMMF.Read(writtenData, 0, -1);

                    //var t3 = Environment.TickCount; if (t3 - t2 > 0) Debug.WriteLine("DESPUES de server.Read {0} - diff: {1}", ID, t3 - t2);

                    if (amount == 0)
                        throw new Exception("Read 0 bytes from queue!");

                    Output.outString("Read data: {0}", BitConverter.ToString(writtenData));

                    // PacketInfo HEADER
                    var packetInfo = new PacketInfo();

                    var piSize = Marshal.SizeOf(packetInfo);
                    var piData = new byte[piSize];

                    // Reads the piSize bytes corresponding to the packetInfo
                    Buffer.BlockCopy(writtenData, 0, piData, 0, piSize);

                    // Converts the bytes to a packetInfo
                    IntPtr ptr = Marshal.AllocHGlobal(piSize);
                    Marshal.Copy(piData, 0, ptr, piSize);
                    packetInfo = (PacketInfo)Marshal.PtrToStructure(ptr, packetInfo.GetType());

                    var cSize = packetInfo.Size;
                    var cData = new byte[cSize];

                    // Packet size can not be bigger than buffer size
                    Debug.Assert(cSize < _pktMmfBufferSize);

                    Buffer.BlockCopy(writtenData, piSize, cData, 0, cSize);

                    PacketManager.PacketReceived(packetInfo, ref cData);

                    //var t4 = Environment.TickCount; if (t4- t3 > 0) Debug.WriteLine("DESPUES de PacketManager.PacketReceived {0} - diff: {1}", ID, t4 - t3);

                }
            };

            // Start reader thread
            ThreadPool.QueueUserWorkItem(o => { reader(); });
             
        }


        // Memory Mapped File - Command MMF data

        private static string _cmdMmfName;
        private static SharedMemory.CircularBuffer cmdMMF;
        public static void WriteCommandToCmdMMF(ServerCodes sc, byte[] data = null, UInt16 newLength = 0, ushort socketId = 0)
        {
            int mmfWriteTimeout = 1000;
            byte[] writtenData;

            if (sc == ServerCodes.SCODE_INJECTPACKET)
            {
                /*
                    |   ServerCode   |  SocketID   |   newLength     |                   Packet                  |
                    |    1-byte      |   2-bytes   |    2-bytes      |---------------     data      -------------|
                    |------------------------------------    writtenData   --------------------------------------|
                */
                writtenData = new byte[1 + 2 + 2 + newLength];
                writtenData[0] = (byte)sc;
                Array.Copy(FilterManager.StructToBytes(socketId), 0, writtenData, 1, 2);
                Array.Copy(FilterManager.StructToBytes(newLength), 0, writtenData, 3, 2);
                Array.Copy(data, 0, writtenData, 5, newLength);
            }
            else if (sc == ServerCodes.SCODE_STARTFILTERING)
            {
                /*
                    |   ServerCode   |  FilterCount   |             Filter Data                     |
                    |    1-byte      |   1-byte       |-----------------Data------------------------|
                    |------------------------   writtenData  ---------------------------------------|
                */
                writtenData = new byte[FilterManager.FILTERINBYTESSIZE * FilterManager.GetActiveFilterCount() + 2]; // Minimo:  SizeOf(Filter) * cantActivos + 2 
                writtenData[0] = (byte)sc;
                writtenData[1] = (byte)FilterManager.GetActiveFilterCount();
                byte[] bytes = FilterManager.ConvertFilterListToBytes();
                Array.Copy(bytes, 0, writtenData, 2, FilterManager.FILTERINBYTESSIZE * FilterManager.GetActiveFilterCount());
            }
            else if (sc == ServerCodes.SCODE_SETPACKET)
            {
                /*
                    |   ServerCode   |  newLength    |             Packet Data                          |
                    |    1-byte      |   2-bytes     |-----------------Data-----------------------------|
                    |-------------------------------------   writtenData    ----------------------------|
                */
                writtenData = new byte[1 + 2 + newLength];
                writtenData[0] = (byte)sc;
                Array.Copy(FilterManager.StructToBytes(newLength), 0, writtenData, 1, 2);
                Array.Copy(data, 0, writtenData, 3, newLength);
            }
            else if (sc == ServerCodes.SCODE_LOADDLLEX)
            {
                /*
                    |   ServerCode   |           fileName string        |
                    |    1-byte      |----------  string.Length --------|
                    |-------------------   writtenData    --------------|
                */
                writtenData = new byte[1 + Settings.DLLEx.Length];
                writtenData[0] = (byte)sc;
                Array.Copy(Encoding.ASCII.GetBytes(Settings.DLLEx), 0, writtenData, 1, Settings.DLLEx.Length);
            }
            else if (sc == ServerCodes.SCODE_STARTCAPTURE || 
                sc == ServerCodes.SCODE_STOPCAPTURE || 
                sc == ServerCodes.SCODE_STOPFILTERING || 
                sc == ServerCodes.SCODE_UNLOADDLLEX)
            {
                writtenData = new byte[1];
                writtenData[0] = (byte)sc;
            }
            else throw new NotImplementedException("ServerCode " + (int)sc + " NOT IMPLEMENTED");

            // Spawneamos un thread que puede esperar el mmfWriteTimeout para escribir el comando
            //int threadCount = 0;

            //int myThreadIndex = Interlocked.Increment(ref threadCount);

            if (CheckMMFileExists(_cmdMmfName))
                cmdMMF = new SharedMemory.CircularBuffer(_cmdMmfName);
            else
            {
                System.Windows.Forms.MessageBox.Show("ERROR, MMF does not exists!"); // No esta el archivo
                return;
            }

            //The number of milliseconds to wait, or Timeout.Infinite (-1) to wait indefinitely. 
            int amount = cmdMMF.Write(writtenData, 0, mmfWriteTimeout);

            if (amount == 0)
                throw new Exception("Write 0 bytes to command MMF!");

            //Output.outString("Write data: {0}", BitConverter.ToString(writtenData));

        }


    }
}
